repondeur/zam_repondeur/resources.py

```diff
@@ -0,0 +1,179 @@
+from typing import Iterator, List, Optional
+
+from pyramid.request import Request
+
+from zam_repondeur.models import (
+    Amendement as AmendementModel,
+    Lecture as LectureModel,
+    DBSession,
+)
+
+
+class Resource(dict):
+    """Location-aware resource"""
+
+    __name__: Optional[str] = None
+    __parent__: Optional["Resource"] = None
+
+    @property
+    def breadcrumbs_title(self) -> str:
+        raise NotImplementedError
+
+    breadcrumbs_link: bool = True
+
+    def __init__(self, name: str, parent: "Resource") -> None:
+        self.__name__ = name
+        self.__parent__ = parent
+
+    @property
+    def parents(self) -> Iterator["Resource"]:
+        parent = self.__parent__
+        while parent is not None:
+            yield parent
+            parent = parent.__parent__
+
+    @property
+    def ancestors(self) -> List["Resource"]:
+        return list(reversed(list(self.parents)))
+
+    def add_child(self, child: "Resource") -> None:
+        self[child.__name__] = child
+
+
+class Root(Resource):
+    def __init__(self, request: Request) -> None:
+        self.add_child(LectureCollection(name="lectures", parent=self))
+
+
+class LectureCollection(Resource):
+
+    breadcrumbs_title = "Lectures"
+
+    def models(self) -> List[LectureModel]:
+        return LectureModel.all()
+
+    def __getitem__(self, key: str) -> Resource:
+        try:
+            chambre, session, num_texte, organe = key.split(".")
+        except ValueError:
+            raise KeyError
+        return LectureResource(
+            name=key,
+            parent=self,
+            chambre=chambre,
+            session=session,
+            num_texte=int(num_texte),
+            organe=organe,
+        )
+
+
+class LectureResource(Resource):
+    def __init__(
+        self,
+        name: str,
+        parent: Resource,
+        chambre: str,
+        session: str,
+        num_texte: int,
+        organe: str,
+    ) -> None:
+        super().__init__(name=name, parent=parent)
+        self.chambre = chambre
+        self.session = session
+        self.num_texte = num_texte
+        self.organe = organe
+        self.add_child(AmendementCollection(name="amendements", parent=self))
+        self.add_child(ArticleCollection(name="articles", parent=self))
+
+    def model(self) -> LectureModel:
```

Comment:
Not fond of `models/model`, proposals: `list/item`? `collection/resource`? (redundant with the name but given the context that might more explicit). Why not properties?

Reply:
Sorry, I don't get it. Can you explain a bit more what you mean here?

Reply:
Oh, got it, nevermind!